midi_utils/notes_counting.py

```diff
@@ -0,0 +1,77 @@
+import numpy as np
+import pandas as pd
+
+
+def aggregate_counts_by_time(record, count_col, time_col):
+    """
+    Utility to aggregate created note counting methods
+    """
+    grouped_record = (
+        record.groupby([time_col])
+        .agg(
+            **{
+                "notes_mean": pd.NamedAgg(count_col, "mean"),
+                "notes_std": pd.NamedAgg(count_col, "std"),
+            }
+        )
+        .reset_index()
+        .assign(
+            lower_bound=lambda x: x["notes_mean"] - x["notes_std"],
+            upper_bound=lambda x: x["notes_mean"] + x["notes_std"],
+        )
+    )
+    # capping count at 0 as we can have negative count
+    grouped_record.loc[:, "lower_bound"] = np.where(grouped_record["lower_bound"] > 0, grouped_record["lower_bound"], 0)
+    return grouped_record
+
+
+def count_notes_lookahead(record: pd.DataFrame, filter_count: int = 2):
+    """
+    Counts how many notes were played simultaneously by looking at notes ahead played in same time interval as initial note
+    """
+    overlapping_indices = []
+    out_record = record.copy()
+
+    for idx_current, row_current in out_record.iterrows():
+        # list of note incdices that are played with the current note
+        notes = [idx_current]
+
+        for idx_next, row_next in out_record.iloc[idx_current + 1 :].iterrows():
+            # gather all notes that start before the current note ends
+            if row_current["end"] > row_next["start"]:
```

Comment:
This is a promising approach to the problem, but with the current implementation I'm seeing a lot of false positives happening on longer notes. It's very common in piano music to hold a note for a couple of seconds, and in that case this includes every note that was played before the current note was released.

In the case from your notebook:

```
In [24]: record_lookahead2[["start", "overlapping_indices"]]
              start                                overlapping_indices
0          1.934219                                          [0, 1, 2]
1          1.935573            [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12]
```

The second sequence is spanning 12 notes and over 5 seconds.